File: frontend/src/lib/api/protocol.ts
import { createApiClient } from "@/lib/api/client";
import type { GovernanceProposal, ProposalLifecycleEvent, ReferralCampaign } from "@/types";

const protocolClient = createApiClient({ basePath: "/protocol" });

interface ApiProposalLifecycleEvent {
  sequence: number;
  from_status?: string | null;
  to_status: string;
  occurred_at: string;
  detail: string;
}

interface ApiGovernanceProposal {
  id: string;
  title: string;
  proposer: string;
  status: GovernanceProposal["status"];
  participation_pct: number;
  executable_after?: string | null;
  lifecycle: ApiProposalLifecycleEvent[];
}

interface ApiReferralReward {
  id: string;
  amount: number;
  status: "pending" | "settled" | "claimed";
}

interface ApiReferralCampaign {
  code: string;
  uses: number;
  rewards_earned: number;
  rewards_pending: number;
  rewards_settled: number;
  rewards_claimed: number;
  conversion_rate_pct: number;
  rewards: ApiReferralReward[];
}

interface ApiReferralSharePayload {
  code: string;
  share_url: string;
  qr_image_base64: string;
}

function mapProposal(proposal: ApiGovernanceProposal): GovernanceProposal {
  return {
    id: proposal.id,
    title: proposal.title,
    proposer: proposal.proposer,
    status: proposal.status,
    participation: `${proposal.participation_pct}%`,
    executableAt: proposal.executable_after
      ? new Date(proposal.executable_after).toLocaleString()
      : "Pending",
    lifecycle: proposal.lifecycle.map(
      (event): ProposalLifecycleEvent => ({
        sequence: event.sequence,
        from_status: event.from_status,
        to_status: event.to_status,
        occurred_at: event.occurred_at,
        detail: event.detail,
      })
    ),
  };
}

function formatReward(amount: number): string {
  return `$${(amount / 100).toFixed(0)}`;
}

function mapReferral(campaign: ApiReferralCampaign, share?: ApiReferralSharePayload): ReferralCampaign {
  return {
    code: campaign.code,
    referrals: campaign.uses,
    rewards: formatReward(campaign.rewards_earned),
    conversionRate: `${campaign.conversion_rate_pct}%`,
    rewardsPending: formatReward(campaign.rewards_pending),
    rewardsSettled: formatReward(campaign.rewards_settled),
    rewardsClaimed: formatReward(campaign.rewards_claimed),
    shareUrl: share?.share_url,
    qrImageBase64: share?.qr_image_base64,
  };
}

export async function listGovernanceProposals(): Promise<GovernanceProposal[]> {
  const proposals = await protocolClient.get<ApiGovernanceProposal[]>("/governance/proposals");
  return proposals.map(mapProposal);
}

export async function listReferralCampaigns(): Promise<ReferralCampaign[]> {
  const campaigns = await protocolClient.get<ApiReferralCampaign[]>("/referrals");
  const enriched = await Promise.all(
    campaigns.map(async (campaign) => {
      try {
        const share = await protocolClient.get<ApiReferralSharePayload>(
          `/referrals/${campaign.code}/share`
        );
        return mapReferral(campaign, share);
      } catch {
        return mapReferral(campaign);
      }
    })
  );
  return enriched;
}

File: backend/app/schemas/protocol.py
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field

ProposalStatus = Literal["active", "succeeded", "defeated", "executed", "cancelled"]
ReferralRewardStatus = Literal["pending", "settled", "claimed"]


class ProposalLifecycleEvent(BaseModel):
    sequence: int
    from_status: Optional[str] = None
    to_status: str
    occurred_at: datetime
    detail: str


class GovernanceProposalCreate(BaseModel):
    title: str = Field(min_length=3, max_length=200)
    description: str = Field(min_length=10, max_length=4000)
    proposer: str
    onchain_id: Optional[int] = None
    voting_ends_at: Optional[datetime] = None
    executable_after: Optional[datetime] = None


class GovernanceProposalResponse(BaseModel):
    id: str
    onchain_id: Optional[int] = None
    title: str
    description: str
    proposer: str
    status: str
    for_votes: int
    against_votes: int
    abstain_votes: int
    participation_pct: float
    voting_ends_at: Optional[datetime] = None
    executable_after: Optional[datetime] = None
    lifecycle: list[ProposalLifecycleEvent] = Field(default_factory=list)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class VotingPowerBreakdown(BaseModel):
    address: str
    self_power: int
    delegated_in: int
    effective_power: int
    has_outbound_delegation: bool


class ReferralCampaignCreate(BaseModel):
    code: str = Field(min_length=4, max_length=32)
    owner: str


class ReferralRewardResponse(BaseModel):
    id: str
    code: str
    swap_id: Optional[int] = None
    amount: int
    status: str
    settled_at: Optional[datetime] = None
    claimed_at: Optional[datetime] = None
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ReferralCampaignResponse(BaseModel):
    id: str
    code: str
    owner: str
    uses: int
    rewards_earned: int
    rewards_pending: int
    rewards_settled: int
    rewards_claimed: int
    conversion_rate_pct: float
    rewards: list[ReferralRewardResponse] = Field(default_factory=list)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ReferralSharePayload(BaseModel):
    code: str
    owner: str
    qr_content: str
    share_url: str
    qr_image_base64: str
    created_at: datetime
